plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,220 @@
+package avro
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jeremywohl/flatten/v2"
+	"github.com/linkedin/goavro/v2"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+// If SchemaRegistry is set, we assume that our input will be in
+// Confluent Wire Format
+// (https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format) and we will load the schema from the registry.
+
+// If Schema is set, we assume the input will be Avro binary format, without
+// an attached schema or schema fingerprint
+
+type Parser struct {
+	MetricName      string            `toml:"metric_name"`
+	SchemaRegistry  string            `toml:"avro_schema_registry"`
+	Schema          string            `toml:"avro_schema"`
+	Measurement     string            `toml:"avro_measurement"`
+	Tags            []string          `toml:"avro_tags"`
+	Fields          []string          `toml:"avro_fields"`
+	Timestamp       string            `toml:"avro_timestamp"`
+	TimestampFormat string            `toml:"avro_timestamp_format"`
+	FieldSeparator  string            `toml:"avro_field_separator"`
+	DefaultTags     map[string]string `toml:"-"`
+
+	Log         telegraf.Logger `toml:"-"`
+	registryObj *schemaRegistry
+}
+
+func (p *Parser) Init() error {
+	if (p.Schema == "" && p.SchemaRegistry == "") || (p.Schema != "" && p.SchemaRegistry != "") {
+		return errors.New("exactly one of 'schema_registry' or 'schema' must be specified")
+	}
+	if p.TimestampFormat == "" {
+		return errors.New("must specify 'timestamp_format'")
+	}
```

Comment:
From the user side, why do they need always need to specify the format? If the timestamp is empty, then the current time is used and we know what to do. Shouldn't the format only be required if the timestamp is not nil?